data/preprocess/util.py

```diff
@@ -0,0 +1,362 @@
+import numpy as np
+from os import path
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, List, Optional, Union, Iterable
+import yaml
+
+### DEFAULT PREPROCESS UTILS ##########################################################
+
+
+@dataclass
+class RawSplitBase:
+    """Base RawSplit, tracks data and meta data of a raw split.
+
+    Attributes:
+        file (str): File with tab-separated raw triples (can be labeled)
+        collect_entities (bool): If true, entities contained in this split will be
+            collected during preprocessing.
+        collect_relations (bool): If true, relations contained in this split will be
+            collected during preprocessing.
+        raw_data (list): List of triples represented with raw ids; determined during
+            preprocessing
+        size (str): Number of triples; determined during preprocessing.
+        entities (dict): Entities contained in this split if collected. Keys refer to
+            raw id's and values to a dense index assigned during pre-processing.
+        relations (dict): See entities.
+
+    """
+
+    # fields defined by user
+    file: str
+    collect_entities: bool = False
+    collect_relations: bool = False
+
+    # fields defined during pre-processing
+    raw_data: List[str] = None
+    order_sop: bool = False
+    size: int = None
+
+    entities: Dict = field(default_factory=dict)
+    relations: Dict = field(default_factory=dict)
+
+    def write_splits(self, entities: Dict, relations: Dict, folder: str):
+        """Write the derived splits of this RawSplit and collect meta-data. """
+        raise NotImplemented
+
+    def update_config(self, config: Dict) -> Dict:
+        """Update a dataset config with meta data of the derived splits."""
+        raise NotImplemented
+
+
+@dataclass
+class RawSplit(RawSplitBase):
+    """A raw split, tracks meta data and info about splits to be derived.
+
+    Attributes:
+        derived_split_key (str): The config key of the default split derived from
+            this raw split.
+        derived_split_options (dict): A dictionary with config options to
+            be added to the config entry of the derived default split.
+        derived_sample_split_key (str): If a key is set, an additional subsample
+            with config entry of this key will be derived from the raw split.
+        sample_split_options (dict): Dict with options to be added to the config
+            entry of the subsample split.
+        sample_size (int): Size of the subsample split.
```

Comment:
As said above: I think there should be a DerivedSplitBase class that can be implemented as FilteredDerivedSplit and SampledDerivedSplit. They should be called in a loop after the main split is finished.